README.md
# ubiquiti-video

This is a command-line tool to pull video from Ubiquiti's UniFi protect into MP4 or MKV files.

NOTE: In this document we assume your DVR's IP address is `192.168.1.1`. If it is different you'll need to adjust the
commands and `.env` file accordingly.

## How to use this script

If you've never run this script before just run `pnpm install` and it should install everything you need.

To fetch video from a camera you'll need the camera's ID. If you don't have your camera IDs you can get them with a
tool like [ubiquiti-db](https://github.com/timmattison/ubiquiti-db).

### Fetching video

Fetching video from a camera can be done like this:

```bash
./ubiquiti-video.ts fetch -i "63eeeeeeefffffffeeeaab68" -s 2023-07-22T15:40:00 -e 2023-07-22T15:42:00
```

This will fetch two minutes of video for the camera with the ID `63eeeeeeefffffffeeeaab68` from July 22nd, 2023 at
15:40:00 to 15:42:00 and convert the video to `.mkv` format.

The file will be called `2023-07-22_03:40:00_PM_2023-07-22_03:42:00_PM_63eeeeeeefffffffeeeaab68.mkv`.

If you want the video to remain in `.mp4` format just add the `-m` flag to the command.

## Setup

To set this script up you'll need to:

- Have ffmpeg installed (if you want `.mkv` format videos)
- Configure the `.env` file values
- Install the dependencies

### Fill in the .env file

This repository contains a file called `.env.example` with some sample values in it. You'll need to copy/move this file
to `.env` and fill in your own values. The example file looks like this:

```text
export UbiquitiUsername=USERNAME
export UbiquitiPassword=PASSWORD
export UbiquitiIp=192.168.1.1
```

Fill in your Ubiquiti credentials in the username and password fields. For some people this will be their ui.com
credentials.
For other people with specific local credentials you'll need to use those instead.

Change the IP address to the IP address of your DVR, if necessary.

### Install the dependencies

This script has only been tested with `pnpm`. `npm` and `yarn` might work but you're on your own.

```bash
pnpm install
```


CHANGELOG.md
# ubiquiti-video

## 2.0.0

### Major Changes

- 8560867: Script no longer uses SSH and Prisma


prisma-client.ts
import { PrismaClient } from '@prisma/client'
import { NodeSSH } from 'node-ssh'
import os from 'os'
import net from 'net'
import { UbiquitiEnvironment } from './ubiquiti-environment.ts'

interface PrismaOverSsh {
  prismaClient: any
  ssh: NodeSSH
  localServer: net.Server
}

export async function getPrismaClientOverSsh() {
  const tunnel = await buildTunnelToNvr()

  const result: PrismaOverSsh = {
    prismaClient: new PrismaClient(),
    ssh: tunnel.ssh,
    localServer: tunnel.localServer,
  }

  return result
}

async function buildTunnelToNvr() {
  const ssh = await new NodeSSH().connect({
    privateKeyPath: `${os.homedir()}/.ssh/id_rsa`,
    host: UbiquitiEnvironment.UbiquitiIp,
    username: UbiquitiEnvironment.UbiquitiSshUsername,
  })

  const localServer = net.createServer((localSocket) => {
    ssh
      .forwardOut(
        'localhost',
        0, // Use a random source port
        '127.0.0.1',
        5433 // destination port (remote server)
      )
      .then((remoteSocket) => {
        // Relay traffic between localSocket and remoteSocket
        localSocket.pipe(remoteSocket).pipe(localSocket)

        remoteSocket.on('error', (error: Error) => {
          console.error('Remote socket error:', error)
        })
      })
      .catch((error) => {
        console.error('Error forwarding SSH connection:', error)
      })
  })

  localServer.listen(5433, '0.0.0.0', () => {
    // console.log('Local server listening on 0.0.0.0:5433')
  })

  localServer.on('error', (error) => {
    // console.error('Local server error:', error)
  })

  // Close the local server and SSH connection on exit
  process.on('SIGINT', () => {
    localServer.close()
    ssh.dispose()
    process.exit()
  })

  // Just in case, this tiny sleep here makes sure that the other tasks get a chance to actually run before we return
  await delay(10)

  return { ssh, localServer }
}

async function delay(ms: number): Promise<void> {
  await new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}


ubiquiti-video.ts
#!/usr/bin/env -S npx ts-node-esm

import { fileURLToPath } from 'url'
import {
  command,
  flag,
  number,
  option,
  optional,
  run,
  string,
  subcommands,
} from 'cmd-ts'
import { z } from 'zod'
import fs from 'fs'
import path from 'path'
import { add, startOfDay } from 'date-fns'
import ffmpeg, { type FfprobeData } from 'fluent-ffmpeg'
import pLimit from 'p-limit'
import { UbiquitiEnvironment } from './ubiquiti-environment.ts'
import ora from 'ora'
import {
  fetchVideo,
  type Status,
  StatusType,
} from '@timmattison/uprotect-fetch'

export function scriptDirname() {
  return path.dirname(fileURLToPath(import.meta.url))
}

const ArgumentsSchema = z.object({
  cameraId: z.string(),
  cameraName: z.string().optional(),
  start: z.date(),
  end: z.date(),
  mp4: z.boolean().default(false),
})

type ArgumentsType = z.infer<typeof ArgumentsSchema>

async function main(args: ArgumentsType) {
  const cameras = [{ id: args.cameraId, name: args.cameraName }]
  const auth = {
    username: UbiquitiEnvironment.UbiquitiUsername,
    password: UbiquitiEnvironment.UbiquitiPassword,
  }

  const spinner = ora().start()

  const result = await fetchVideo({
    ipAddress: UbiquitiEnvironment.UbiquitiIp,
    auth,
    cameras,
    start: args.start,
    end: args.end,
    mp4: args.mp4,
    statusCallback: (status: Status) => {
      switch (status.type) {
        case StatusType.Waiting:
          spinner.text = 'Waiting to start downloading...'
          break
        case StatusType.Downloading:
          if (status.progressPercent !== undefined) {
            spinner.text = `Downloading: ${String(status.progressPercent)}`
          } else {
            spinner.text = 'Downloading...'
          }
          break
        case StatusType.DownloadThroughput:
          if (status.throughputString !== undefined) {
            spinner.text = `Download throughput: ${String(
              status.throughputString,
            )}/s`
          }
          break
        case StatusType.Converting:
          spinner.text = `Converting...`
          break
        case StatusType.Done:
          if (status.filename !== undefined) {
            spinner.succeed(`Done [${String(status.filename)}]`)
          } else {
            spinner.succeed('Done')
          }
          spinner.stop()
          break
        case StatusType.AllDone:
          spinner.stop()
          break
        case StatusType.Error:
          throw status.error
      }
    },
  })
  JSON.stringify(result, null, 2)
}

const NameVersionDescriptionSchema = z.object({
  name: z.string(),
  description: z.string(),
  version: z.string(),
})

const packageInfo = NameVersionDescriptionSchema.parse(
  JSON.parse(
    fs.readFileSync(path.resolve(scriptDirname(), './package.json'), 'utf8'),
  ),
)

const StartOfDaySchema = z
  .string()
  .regex(/^[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}$/)
  .transform((data) => {
    const splitInput = data.split('-')
    const year = Number(splitInput[0])
    const month = Number(splitInput[1])
    const day = Number(splitInput[2])
    return startOfDay(new Date(year, month - 1, day))
  })

const fetchDayCommand = command({
  name: 'fetch-day',
  args: {
    cameraId: option({
      long: 'camera-id',
      short: 'i',
      type: string,
      description: 'ID of the camera',
    }),
    cameraName: option({
      long: 'camera-name',
      short: 'n',
      type: optional(string),
      description: 'name of the camera (for naming output files)',
    }),
    date: option({
      long: 'date',
      short: 's',
      type: string,
      description: 'date (formatted as YYYY-MM-DD)',
    }),
    mp4: flag({
      long: 'mp4',
      short: 'm',
      description: 'keep as mp4 (otherwise convert to mkv for compatibility)',
      defaultValue: () => false,
    }),
  },
  handler: (args) => {
    const start = StartOfDaySchema.parse(args.date)
    const end = add(start, { hours: 24 })
    void main(
      ArgumentsSchema.parse({
        ...args,
        start,
        end,
      }),
    ).finally()
  },
})

const fetchCommand = command({
  name: 'fetch',
  args: {
    cameraId: option({
      long: 'camera-id',
      short: 'i',
      type: string,
      description: 'ID of the camera',
    }),
    cameraName: option({
      long: 'camera-name',
      short: 'n',
      type: optional(string),
      description: 'name of the camera (for naming output files)',
    }),
    start: option({
      long: 'start',
      short: 's',
      type: string,
      description:
        'start time (formatted as ISO 8601 - e.g. "2011-10-05T14:48:00.000Z")',
    }),
    end: option({
      long: 'end',
      short: 'e',
      type: string,
      description:
        'end time (formatted as ISO 8601 - e.g. "2011-10-05T14:49:00.000Z")',
    }),
    mp4: flag({
      long: 'mp4',
      short: 'm',
      description: 'keep as mp4 (otherwise convert to mkv for compatibility)',
      defaultValue: () => false,
    }),
  },
  handler: (args) => {
    void main(
      ArgumentsSchema.parse({
        ...args,
        start: new Date(args.start),
        end: new Date(args.end),
      }),
    ).finally()
  },
})

const defaultThreads = 10
const extractAudioCommand = command({
  name: 'extract-audio',
  args: {
    inputFile: option({
      long: 'input-file',
      short: 'i',
      type: string,
      description: 'the input file',
    }),
    chunkDuration: option({
      long: 'chunk-duration',
      short: 'd',
      type: number,
      description: 'the duration of the chunks of audio extracted (in seconds)',
    }),
    threads: option({
      long: 'threads',
      short: 't',
      type: number,
      description: `the number of concurrent threads to use to process the audio (default: ${defaultThreads})`,
      defaultValue: () => defaultThreads,
    }),
  },
  handler: async (args) => {
    const chunkDuration = args.chunkDuration
    const limit = pLimit(args.threads)

    const durationInSeconds = await getInputDuration(args.inputFile)
    const numberOfChunks = durationInSeconds - chunkDuration

    try {
      const promises: Array<Promise<string>> = []
      for (let loop = 0; loop < numberOfChunks; loop++) {
        promises.push(
          limit(
            async () => await extractAudio(args.inputFile, loop, chunkDuration),
          ),
        )
      }
      const result = await Promise.all(promises)
      console.log(result)
    } catch (error) {
      console.log('Processing finished with error:', error)
    }
  },
})

async function getInputDuration(inputFile: string) {
  return await new Promise<number>((resolve, reject) => {
    ffmpeg(inputFile).ffprobe((err: Error | null, metadata: FfprobeData) => {
      if (err != null) {
        reject(err)
        return
      }

      if (metadata.format.duration === undefined) {
        reject(
          new Error(`The duration of ${inputFile} could not be determined`),
        )
        return
      }

      resolve(metadata.format.duration)
    })
  })
}

async function extractAudio(
  inputFile: string,
  startTime: number,
  chunkDuration: number,
) {
  const outputFormat = 'wav'
  const startMinute = Math.floor(startTime / 60)
  const startSecond = Math.floor(startTime % 60)
  const startString = `${String(startMinute).padStart(2, '0')}m-${String(
    startSecond,
  ).padStart(2, '0')}s`
  const outputFile = `${inputFile}-audio-${startString}.${outputFormat}`

  return await new Promise<string>((resolve, reject) => {
    ffmpeg(inputFile)
      .setStartTime(startTime)
      .duration(chunkDuration)
      // .audioCodec('pcm_s16le')
      .toFormat(outputFormat)
      .on('end', () => {
        console.log(`Segment ${startTime} processed`)
        resolve(outputFile)
      })
      .on('error', (err) => {
        console.error(err.message)
        reject(err)
      })
      // Save audio file named by its order.
      .save(outputFile)
  })
}

const commandLineParser = subcommands({
  name: packageInfo.name,
  description: packageInfo.description,
  version: packageInfo.version,
  cmds: {
    fetch: fetchCommand,
    'fetch-day': fetchDayCommand,
    'extract-audio': extractAudioCommand,
  },
})

void run(commandLineParser, process.argv.slice(2)).finally()


ubiquiti.ts
import { z } from 'zod'
import { URLSearchParams } from 'url'
import axios, { type AxiosResponse } from 'axios'
import axiosRetry from 'axios-retry'
import * as https from 'https'
import * as fs from 'fs'
import { UbiquitiEnvironment } from './ubiquiti-environment.ts'
import { getPrismaClientOverSsh } from './prisma-client.ts'
import ffmpeg from 'fluent-ffmpeg'
import {
  add,
  differenceInMinutes,
  differenceInSeconds,
  format,
  formatDistanceStrict,
  isBefore,
} from 'date-fns'
import ora from 'ora'

export async function getFullCameras() {
  const prismaClientOverSsh = await getPrismaClientOverSsh()

  const cameras = await prismaClientOverSsh.prismaClient.cameras.findMany()

  await prismaClientOverSsh.prismaClient.$disconnect()
  prismaClientOverSsh.localServer.close()
  prismaClientOverSsh.ssh.dispose()

  return UbiquitiCamerasSchema.parse(cameras)
}

export async function getSimpleCameras() {
  return UbiquitiCamerasSchema.parse(await getFullCameras()).map((value) => ({
    id: value.id,
    name: value.name,
  }))
}

const UbiquitiCameraSchema = z.object({
  id: z.string().min(24).max(24),
  mac: z.string().min(12).max(12),
  host: z.string().ip(),
  connectionHost: z.string().ip(),
  type: z.string(),
  name: z.string(),
  channels: z.array(
    z.object({
      enabled: z.boolean(),
      isRtspEnabled: z.boolean(),
      width: z.number(),
      height: z.number(),
      fps: z.number(),
      bitrate: z.number(),
      minBitrate: z.number(),
      maxBitrate: z.number(),
    }),
  ),
  stats: z.object({
    rxBytes: z.number(),
    txBytes: z.number(),
    wifi: z.object({
      channel: z.any().optional(),
      frequency: z.any().optional(),
      linkSpeedMbps: z.any().optional(),
      signalQuality: z.number(),
      signalStrength: z.number(),
    }),
    battery: z.object({
      percentage: z.any().optional(),
      isCharging: z.boolean(),
    }),
    video: z.object({
      recordingStart: z.number().nullable(),
      recordingEnd: z.number().nullable(),
      recordingStartLQ: z.number().nullable(),
      recordingEndLQ: z.number().nullable(),
      timelapseStart: z.number().nullable(),
      timelapseEnd: z.number().nullable(),
      timelapseStartLQ: z.number().nullable(),
      timelapseEndLQ: z.number().nullable(),
    }),
  }),
})

const UbiquitiCamerasSchema = z.array(UbiquitiCameraSchema)

type UbiquitiCamerasType = z.infer<typeof UbiquitiCamerasSchema>

type Payload = Record<string, any>
type Cookies = Record<string, string>

async function postWithCookies(
  url: string,
  payload: Payload,
  cookies: Cookies = {},
): Promise<{
  responseText: string
  updatedCookies: Cookies
}> {
  const headers = {
    'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
    Cookie: Object.entries(cookies)
      .map(([key, value]) => `${key}=${value}`)
      .join('; '),
  }

  const body = new URLSearchParams(payload).toString()

  // Configure retries
  axiosRetry(axios, {
    retries: 3,
    // Retry on Network Errors & 4xx responses
    retryCondition: (error) => {
      return error.response == null || error.response.status >= 400
    },
  })

  const response: AxiosResponse = await axios.post(url, body, {
    headers,
    httpsAgent: new https.Agent({
      rejectUnauthorized: false,
    }),
    withCredentials: true,
  })

  const responseText = response.data

  const setCookieHeader = response.headers['set-cookie']
  const updatedCookies = { ...cookies }

  if (setCookieHeader !== undefined) {
    setCookieHeader.forEach((cookie: string) => {
      const [key, value] = cookie.split(';')[0].split('=')
      updatedCookies[key.trim()] = value.trim()
    })
  }

  return { responseText, updatedCookies }
}

function percent(input: number | undefined) {
  if (input === undefined) return 'Unknown'

  return input.toLocaleString(undefined, {
    style: 'percent',
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  })
}

async function getBinaryDataWithCookies(
  url: string,
  cookies: Cookies = {},
  filename: string,
  logMessage: string,
): Promise<{
  updatedCookies: Cookies
}> {
  const headers = {
    Cookie: Object.entries(cookies)
      .map(([key, value]) => `${key}=${value}`)
      .join('; '),
  }

  const writer = fs.createWriteStream(filename)

  const spinner = ora(`Waiting to download... ${logMessage}`).start()

  const response: AxiosResponse = await axios.get(url, {
    headers,
    httpsAgent: new https.Agent({
      rejectUnauthorized: false,
    }),
    withCredentials: true,
    responseType: 'stream',
    onDownloadProgress: (progressEvent) => {
      spinner.color = 'yellow'
      spinner.text = `Downloading... ${percent(
        progressEvent.progress,
      )} ${logMessage}`
    },
  })

  response.data.pipe(writer)

  const setCookieHeader = response.headers['set-cookie']
  const updatedCookies = { ...cookies }

  if (setCookieHeader != null) {
    setCookieHeader.forEach((cookie: string) => {
      const [key, value] = cookie.split(';')[0].split('=')
      updatedCookies[key.trim()] = value.trim()
    })
  }

  return await new Promise<{ updatedCookies: Cookies }>((resolve, reject) => {
    writer.on('finish', () => {
      spinner.color = 'green'
      spinner.text = `Download complete of ${logMessage}`
      spinner.stop()
      resolve({ updatedCookies })
    })
    writer.on('error', (err) => {
      spinner.fail('Error writing to the file')
      reject(err)
    })
    response.data.on('error', (err: Error) => {
      spinner.fail('Error reading the response from the server')
      reject(err)
    })
  })
}

const auth: Payload = {
  username: UbiquitiEnvironment.UbiquitiUsername,
  password: UbiquitiEnvironment.UbiquitiPassword,
}

export interface StartAndEndStrings {
  start: string
  end: string
}

export interface StartAndEndEpochs {
  start: number
  end: number
}

async function getCookies() {
  const cookies: Cookies = {}

  // const { responseText: loginResult, updatedCookies: loginCookies } =
  const { updatedCookies: loginCookies } = await postWithCookies(
    'https://192.168.0.1:443/api/auth/login',
    auth,
    cookies,
  )

  return loginCookies
}

const OutputFileSchema = z.object({
  camera: UbiquitiCameraSchema.pick({ id: true, name: true }),
  start: z.date(),
  end: z.date(),
  filename: z.string(),
})

type OutputFileType = z.infer<typeof OutputFileSchema>

async function convertToMkv(inputPath: string, outputPath: string) {
  return await new Promise((resolve, reject) => {
    ffmpeg(inputPath)
      .outputOptions('-c', 'copy') // This copies the existing audio and video streams without re-encoding
      .output(outputPath)
      .on('start', function () {
        // console.log('Conversion started')
      })
      .on('progress', function (progress) {
        // console.log(`Progress: ${progress.percent}% done`)
      })
      .on('end', function () {
        // console.log('Conversion ended successfully')
        resolve(true)
      })
      .on('error', function (err) {
        console.log('Conversion error: ', err)
        reject(err)
      })
      .run()
  })
}

function formatDate(date: Date) {
  return format(date, 'yyyy-MM-dd hh:mm:ss aa')
}

export async function getVideo(
  cameras: UbiquitiCamerasType,
  start: Date,
  end: Date,
  mp4: boolean,
) {
  const outputFiles: OutputFileType[] = []

  let currentStart = start

  const maxMinuteDiff = {
    minutes: 60,
  }

  while (isBefore(currentStart, end)) {
    const diffInMinutes = differenceInMinutes(end, currentStart)
    let currentEnd: Date

    if (diffInMinutes > maxMinuteDiff.minutes) {
      currentEnd = add(currentStart, maxMinuteDiff)
    } else {
      currentEnd = end
    }

    for (const camera of cameras) {
      const url = `https://192.168.0.1:443/proxy/protect/api/video/export?camera=${
        camera.id
      }&start=${currentStart.getTime()}&end=${currentEnd.getTime()}`

      const formattedStart = formatDate(currentStart)
      const formattedEnd = formatDate(currentEnd)

      const logMessage = `${formattedStart} -> ${formattedEnd} - ${camera.name}`

      const baseFilename = `${formattedStart}_${formattedEnd}_${camera.name}`
      const mp4Filename = `${baseFilename}.mp4`
      const mkvFilename = `${baseFilename}.mkv`

      currentStart = add(currentStart, maxMinuteDiff)

      if (fs.existsSync(mp4Filename)) {
        if (fs.statSync(mp4Filename).size === 0) {
          fs.unlinkSync(mp4Filename)
        } else {
          console.log(
            `${mp4Filename} already exists and is not empty, refusing to overwrite it`,
          )
          continue
        }
      }

      const downloadStart = Date.now()

      // Get cookies each time so we don't time out our login on requests for large videos
      const cookies = await getCookies()

      await getBinaryDataWithCookies(url, cookies, mp4Filename, logMessage)

      const downloadEnd = Date.now()
      const diffInSeconds = differenceInSeconds(downloadEnd, downloadStart)
      const mp4FileStats = fs.statSync(mp4Filename)
      const throughput = mp4FileStats.size / diffInSeconds

      let throughputString: string

      if (throughput > 1000000) {
        throughputString = `${Math.round(throughput / 1000000)} MB`
      } else {
        throughputString = `${Math.round(throughput)} B`
      }

      // Must log something here or the user will only see the live download
      //   updates but won't know how many files were downloaded already
      console.log(
        `Downloaded ${logMessage} in ${formatDistanceStrict(
          downloadEnd,
          downloadStart,
        )} [${throughputString}/s]`,
      )

      if (!mp4) {
        await convertToMkv(mp4Filename, mkvFilename)
        fs.unlinkSync(mp4Filename)
      }

      outputFiles.push(
        OutputFileSchema.parse({
          camera,
          start,
          end,
          filename: mp4 ? mp4Filename : mkvFilename,
        }),
      )
    }
  }

  return outputFiles
}
